lib/wallet/src/Cardano/Wallet/DB.hs

```diff
@@ -406,24 +409,39 @@ mkDBLayerFromParts DBLayerCollection{..} = DBLayer
     , listCheckpoints = listCheckpoints_ dbCheckpoints
     , putWalletMeta = putWalletMeta_ dbWalletMeta
     , readWalletMeta = readWalletMeta_ dbWalletMeta
-    , isStakeKeyRegistered = isStakeKeyRegistered_ dbDelegation
-    , putDelegationCertificate = putDelegationCertificate_ dbDelegation
-    , putDelegationRewardBalance = putDelegationRewardBalance_ dbDelegation
-    , readDelegationRewardBalance = readDelegationRewardBalance_ dbDelegation
+    , isStakeKeyRegistered = \wid -> wrapNoSuchWallet wid $
+        isStakeKeyRegistered_ (dbDelegation wid)
+    , putDelegationCertificate = \wid a b -> wrapNoSuchWallet wid $
+        putDelegationCertificate_ (dbDelegation wid) a b
+    , putDelegationRewardBalance = \wid a -> wrapNoSuchWallet wid $
+        putDelegationRewardBalance_ (dbDelegation wid) a
+    , readDelegationRewardBalance = \wid ->
+        readDelegationRewardBalance_ (dbDelegation wid)
     , putTxHistory = putTxHistory_ dbTxHistory
     , readTxHistory = readTxHistory_ dbTxHistory
     , getTx = getTx_ dbTxHistory
     , putLocalTxSubmission = putLocalTxSubmission_ dbPendingTxs
     , readLocalTxSubmissionPending = readLocalTxSubmissionPending_ dbPendingTxs
     , updatePendingTxForExpiry = updatePendingTxForExpiry_ dbPendingTxs
     , removePendingOrExpiredTx = removePendingOrExpiredTx_ dbPendingTxs
-    , putPrivateKey = putPrivateKey_ dbPrivateKey
-    , readPrivateKey = readPrivateKey_ dbPrivateKey
+    , putPrivateKey = \wid a -> wrapNoSuchWallet wid $
+        putPrivateKey_ (dbPrivateKey wid) a
+    , readPrivateKey = \wid ->
+        readPrivateKey_ (dbPrivateKey wid)
     , readGenesisParameters = readGenesisParameters_ dbWallets
     , rollbackTo = rollbackTo_
     , prune = prune_
     , atomically = atomically_
     }
+  where
+    wrapNoSuchWallet
+        :: WalletId
+        -> stm a
+        -> ExceptT ErrNoSuchWallet stm a
+    wrapNoSuchWallet wid action = ExceptT $ do
+        hasWallet_ dbWallets wid >>= \case
+            False -> pure $ Left $ ErrNoSuchWallet wid
+            True  -> Right <$> action
```

Comment:
I _sometimes_ use the `bool` function to simplify matching on `Bool`:

```suggestion
        hasWallet_ dbWallets wid >>= bool
            (pure $ Left $ ErrNoSuchWallet wid)
            (Right <$> action)
```
(where `bool` is from `Data.Bool`, akin to `maybe` and `either`.)

Though one disadvantage is that it's easy to accidentally switch the argument order, and the compiler will not complain.

Reply:
I'm going to call the `bool`ice. 👮‍♂️

The pattern match is easier to follow indeed, as it does indicate which case is which — and that's pretty much the main source of mistakes when dealing with `Bool`.